caffe2/core/tensor_impl.h

```diff
@@ -524,13 +528,14 @@ class CAFFE2_API TensorImpl : public c10::intrusive_ptr_target {
   inline void* raw_mutable_data(const TypeMeta& meta) {
     // For 0-size tensors it's fine to return any pointer (including nullptr)
     if (storage_.dtype() == meta && (storage_.data() || numel_ == 0)) {
-      return storage_.data();
+      return static_cast<void*>(static_cast<char*>(storage_.data()) + storage_offset_ * meta.itemsize());
     } else {
       CAFFE_ENFORCE_WITH_CALLER(
           numel_ >= 0,
           "Tensor is not initialized. You probably need to call Resize() "
           "before calling mutable_data()");
       bool had_special_dtor = storage_.dtype().dtor() != nullptr;
+      storage_offset_ = 0;
       if (storage_.unique()) {
         storage_.set_dtype(meta);
```

Comment:
does this actually reallocate? (comment below says "because we just reallocated").

Reply:
Sometimes it doesn't "reallocate" (it reuses the old storage) but the net effect is still one of deallocate/reallocate (it's just an optimization). Basically, the first branch is the fast (happy) path.

Reply:
ya, that's what I expected.  I don't know if there's a more accurate way of writing the comment maybe "logically reallocated"or something.